inference-engine/tests/functional/plugin/shared/include/behavior/infer_request_cancellation.hpp

```diff
@@ -0,0 +1,135 @@
+    // Create CNNNetwork from ngrpah::Function
+    InferenceEngine::CNNNetwork cnnNet(function);
+    // Load CNNNetwork to target plugins
+    auto execNet = ie->LoadNetwork(cnnNet, targetDevice, configuration);
+    // Create InferRequest
+    InferenceEngine::InferRequest req = execNet.CreateInferRequest();
+
+    req.StartAsync();
+    auto cancel = std::async(std::launch::async, [&req]{ return req.Cancel(); });
+    InferenceEngine::StatusCode waitStatus = req.Wait(InferenceEngine::IInferRequest::WaitMode::RESULT_READY);
+    InferenceEngine::StatusCode cancelStatus = cancel.get();
+
+    if (targetDevice == CommonTestUtils::DEVICE_CPU) {
+        ASSERT_EQ(static_cast<int>(InferenceEngine::StatusCode::OK), cancelStatus);
+        ASSERT_EQ(static_cast<int>(InferenceEngine::StatusCode::INFER_CANCELLED), waitStatus);
+    } else {
+        ASSERT_EQ(static_cast<int>(InferenceEngine::StatusCode::NOT_IMPLEMENTED), cancelStatus);
+        ASSERT_EQ(static_cast<int>(InferenceEngine::StatusCode::OK), waitStatus);
+    }
+
+    req.StartAsync();
+    waitStatus = req.Wait(InferenceEngine::IInferRequest::WaitMode::RESULT_READY);
+
+    ASSERT_EQ(static_cast<int>(InferenceEngine::StatusCode::OK), waitStatus);
+}
+
+TEST_P(CancellationTests, canCancelBeforeAsyncRequest) {
+    // Skip test according to plugin specific disabledTestPatterns() (if any)
+    SKIP_IF_CURRENT_TEST_IS_DISABLED()
+    // Create CNNNetwork from ngrpah::Function
+    InferenceEngine::CNNNetwork cnnNet(function);
+    // Load CNNNetwork to target plugins
+    auto execNet = ie->LoadNetwork(cnnNet, targetDevice, configuration);
+    // Create InferRequest
+    InferenceEngine::InferRequest req = execNet.CreateInferRequest();
+
+    InferenceEngine::StatusCode cancelStatus = req.Cancel();
+
+    if (targetDevice == CommonTestUtils::DEVICE_CPU) {
+        ASSERT_EQ(static_cast<int>(InferenceEngine::StatusCode::INFER_NOT_STARTED), cancelStatus);
+    } else {
+        ASSERT_EQ(static_cast<int>(InferenceEngine::StatusCode::NOT_IMPLEMENTED), cancelStatus);
+    }
+}
+
+TEST_P(CancellationTests, canCancelInferRequest) {
+    // Skip test according to plugin specific disabledTestPatterns() (if any)
+    SKIP_IF_CURRENT_TEST_IS_DISABLED()
+    // Create CNNNetwork from ngrpah::Function
+    InferenceEngine::CNNNetwork cnnNet(function);
+    // Load CNNNetwork to target plugins
+    auto execNet = ie->LoadNetwork(cnnNet, targetDevice, configuration);
+    // Create InferRequest
+    InferenceEngine::InferRequest req = execNet.CreateInferRequest();
+
+    auto infer = std::async(std::launch::async, [&req]{ req.Infer(); });
+    infer.wait_for(std::chrono::milliseconds(1));
+    InferenceEngine::StatusCode cancelStatus = req.Cancel();
+
+    if (targetDevice == CommonTestUtils::DEVICE_CPU) {
+        ASSERT_EQ(static_cast<int>(InferenceEngine::StatusCode::OK), cancelStatus);
+    } else {
+        ASSERT_EQ(static_cast<int>(InferenceEngine::StatusCode::OK), cancelStatus);
```

Comment:
Are you sure on the _else_ 's assertion (it is same as _if_)?
Again, here the inference may complete BEFORE the CANCELATION  (esp with 1 ms delaying).
so the right check is (InferenceEngine::StatusCode::OK==cancelStatus  AND callback_was_NOT_executed) _OR_ (InferenceEngine::StatusCode::INFER_NOTSTARTED==cancelStatus AND callback_was_executed<so that we know inference has completed>)

Reply:
Right, going to fix that as well.

Reply:
you may populate the timestamp in the callback for the request, and compare that to the current timestamp (populated before calling the Cancel). If timestamp_callback >timestamp_cancel, the Cancel should return OK, otheriwse NOT_STARTD